src/java.base/share/classes/java/util/HexFormat.java

```diff
@@ -0,0 +1,1023 @@
+        }
+        try {
+            return jla.newStringNoRepl(rep, StandardCharsets.ISO_8859_1);
+        } catch (CharacterCodingException cce) {
+            throw new AssertionError(cce);
+        }
+    }
+
+    /**
+     * Returns a byte array containing the parsed hex digits.
+     * A valid string consists only of an even number of hex digits.
+     *
+     * @param string a string containing an even number of only hex digits
+     * @return a byte array
+     * @throws IllegalArgumentException if the string length is not valid or
+     *          the string contains non-hexadecimal characters
+     */
+    private byte[] parseNoDelimiter(CharSequence string) {
+        if ((string.length() & 1) != 0)
+            throw new IllegalArgumentException("string length not even: " +
+                    string.length());
+
+        byte[] bytes = new byte[string.length() / 2];
+        int illegal = 0;        // Accumulate logical-or of all bytes
+        for (int i = 0; i < bytes.length; i++) {
+            int v = fromHexDigits(string, i * 2);
+            bytes[i] = (byte) v;
+            illegal |= v;
+        }
+        // check if any character was an illegal character
+        if (illegal < 0)
+            throw new IllegalArgumentException("input contains non-hexadecimal characters");
+
+        return bytes;
+    }
+
+    /**
+     * Check the number of requested digits against a limit.
+     *
+     * @param fromIndex the initial index of the range, inclusive
+     * @param toIndex the final index of the range, exclusive.
+     * @param limit the maximum allowed
+     * @return the length of the range
+     */
+    private static int checkDigitCount(int fromIndex, int toIndex, int limit) {
+        int length = toIndex - fromIndex;
+        if (length > limit)
+            throw new IllegalArgumentException("string length greater than " +
+                    limit + ": " + length);
+        return length;
+    }
+
+    /**
+     * Returns {@code true} if the character is a valid hexadecimal character or codepoint.
+     * A character is a valid hexadecimal character if
+     * {@link Character#digit(int, int) Character.digit(int, 16)} returns
+     * a positive value.
+     *
+     * @param ch a codepoint
+     * @return {@code true} if the character is valid a hexadecimal character,
+     *          otherwise {@code false}
+     */
+    public boolean isHexDigit(int ch) {
+        return Character.digit(ch, 16) >= 0;
+    }
+
+    /**
+     * Returns the value for the hexadecimal character or codepoint.
+     * The characters {@code "0-9", "A-F", "a-f"} are parsed
+     * using {@link Character#digit(int, int) Character.digit(int, 16)}.
+     *
+     * @param ch a character or codepoint
+     * @return the value {@code 0..15}
+     * @throws  NumberFormatException if the codepoint is not a hexadecimal character
+     */
+    public int fromHexDigit(int ch) {
+        int value = Character.digit(ch, 16);
```

Comment:
Do we need to limit parsing the hex digit for only [0-9a-fA-F]? This would return `0` for other digits, say `fullwidth digit zero` (U+FF10)

Reply:
The normal and conventional characters for hex encoding are limited to the ASCII/Latin1 range.
I don't know of any use case that would take advantage of non-ASCII characters.

Reply:
My point is that probably we should define `hexadecimal string` more clearly. In the class description, that exclusively means [0-9a-fA-F] in the context of formatting, but in the parsing, it allows non-ASCII digits. e.g.,
```
HexFormat.of().parseHex("\uff10\uff11")
```
Succeeds. I would like consistency here.